// @fluentui/react-charting/lib-commonjs/components/TreeChart/TreeChart.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var tslib_1 = require("tslib");
var getStyles = function (props) {
    return {
        root: [
            props.theme.spacing,
            'ms-TreeChart',
            props.className,
            {
                overflow: 'scroll',
            },
        ],
        link: {
            fill: 'none',
            stroke: '#A1A1A1',
            strokeWidth: '2px',
        },
        rectNode: {
            strokeWidth: '2px',
            fill: 'white',
            padding: '10px',
            rx: '2px',
        },
        rectText: tslib_1.__assign({ fill: 'black' }, props.theme.fonts.large),
        rectSubText: tslib_1.__assign({ fill: '#484644' }, props.theme.fonts.medium),
        rectBodyText: tslib_1.__assign({ fill: '#808080' }, props.theme.fonts.xSmall),
        rectMetricText: tslib_1.__assign({ fill: '#000000' }, props.theme.fonts.xLarge),
    };
};
exports.getStyles = getStyles;
