// @fluentui/react-charting/lib-amd/components/PieChart/PieChart.styles.js
define(["require", "exports", "tslib"], function (require, exports, tslib_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    var getStyles = function (props) {
        var className = props.className, theme = props.theme, width = props.width, height = props.height;
        var chartWidth = width + 30;
        var chartPadding = 20;
        var chartHeight = height + 10;
        return {
            root: [
                theme.fonts.medium,
                'ms-PieChart',
                className,
                {
                    width: chartWidth + 2 * chartPadding,
                },
            ],
            chart: [
                {
                    padding: chartPadding,
                    width: chartWidth,
                    height: chartHeight,
                    boxSizing: 'content-box',
                },
            ],
            chartTitle: [
                tslib_1.__assign({ textAlign: 'center' }, theme.fonts.mediumPlus),
            ],
        };
    };
    exports.getStyles = getStyles;
});
