// @fluentui/react-charting/lib/components/HorizontalBarChart/HorizontalBarChart.styles.js
import { __assign } from "tslib";
import { FontSizes, FontWeights } from '@fluentui/react/lib/Styling';
export var getHorizontalBarChartStyles = function (props) {
    var className = props.className, theme = props.theme, width = props.width, barHeight = props.barHeight, showTriangle = props.showTriangle;
    var palette = theme.palette;
    return {
        root: [
            {
                display: 'flex',
                flexDirection: 'column',
                width: width ? width : '100%',
            },
            className,
        ],
        items: {
            marginBottom: showTriangle ? '16px' : '10px',
        },
        chart: {
            width: '100%',
            height: barHeight ? barHeight : 12,
            display: 'block',
        },
        barWrapper: {
            stroke: theme.palette.white,
            strokeWidth: 2,
        },
        chartTitle: __assign(__assign({}, theme.fonts.small), { display: 'flex', justifyContent: 'space-between' }),
        chartTitleLeft: {
            textOverflow: 'ellipsis',
            overflow: 'hidden',
            whiteSpace: 'nowrap',
            display: 'block',
            color: '#171717',
            marginBottom: '5px',
        },
        chartTitleRight: {
            fontSize: FontSizes.medium,
            fontWeight: FontWeights.semibold,
            color: theme.palette.neutralPrimary,
        },
        chartDataTextDenominator: {
            fontSize: FontSizes.medium,
            color: theme.palette.neutralPrimary,
        },
        benchmarkContainer: {
            position: 'relative',
            height: '7px',
            marginTop: '-3px',
            marginBottom: '-1px',
        },
        triangle: {
            width: '0',
            height: '0',
            borderLeft: '4px solid transparent',
            borderRight: '4px solid transparent',
            borderTop: '7px solid',
            borderTopColor: palette.blue,
            marginBottom: '4px',
            position: 'absolute',
        },
    };
};
