// @fluentui/react-charting/lib-amd/components/TreeChart/TreeChart.types.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TreeTraverse = exports.NodesComposition = void 0;
    var NodesComposition;
    (function (NodesComposition) {
        /**
         * NodeComposition enum val for long: number = 1
         */
        NodesComposition[NodesComposition["long"] = 1] = "long";
        /**
         * NodeComposition enum val for compact: number = 0
         */
        NodesComposition[NodesComposition["compact"] = 0] = "compact";
    })(NodesComposition = exports.NodesComposition || (exports.NodesComposition = {}));
    var TreeTraverse;
    (function (TreeTraverse) {
        /**
         * TreeTraverse enum val for preOrder: number = 1
         */
        TreeTraverse[TreeTraverse["preOrder"] = 1] = "preOrder";
        /**
         * TreeTraverse enum val for levelOrder: number = 0
         */
        TreeTraverse[TreeTraverse["levelOrder"] = 0] = "levelOrder";
    })(TreeTraverse = exports.TreeTraverse || (exports.TreeTraverse = {}));
});
