// @fluentui/react-charting/lib/components/DonutChart/Arc/Arc.styles.js
import { DefaultPalette, FontSizes, HighContrastSelectorBlack } from '@fluentui/react/lib/Styling';
export var getStyles = function (props) {
    var _a;
    var color = props.color, href = props.href, theme = props.theme;
    return {
        root: {
            fill: color,
            cursor: href ? 'pointer' : 'default',
            stroke: DefaultPalette.white,
            outline: 'transparent',
            selectors: {
                '::-moz-focus-inner': {
                    border: '0',
                },
            },
        },
        focusRing: {
            stroke: theme.semanticColors.focusBorder,
            strokeWidth: 4,
            fill: 'transparent',
        },
        insideDonutString: {
            fontSize: FontSizes.large,
            fill: theme.semanticColors.bodyText,
            selectors: (_a = {},
                _a[HighContrastSelectorBlack] = {
                    fill: 'rgb(179, 179, 179)',
                },
                _a),
        },
    };
};
