// @fluentui/react-charting/lib-commonjs/components/DonutChart/Arc/Arc.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStyles = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
var getStyles = function (props) {
    var _a;
    var color = props.color, href = props.href, theme = props.theme;
    return {
        root: {
            fill: color,
            cursor: href ? 'pointer' : 'default',
            stroke: Styling_1.DefaultPalette.white,
            outline: 'transparent',
            selectors: {
                '::-moz-focus-inner': {
                    border: '0',
                },
            },
        },
        focusRing: {
            stroke: theme.semanticColors.focusBorder,
            strokeWidth: 4,
            fill: 'transparent',
        },
        insideDonutString: {
            fontSize: Styling_1.FontSizes.large,
            fill: theme.semanticColors.bodyText,
            selectors: (_a = {},
                _a[Styling_1.HighContrastSelectorBlack] = {
                    fill: 'rgb(179, 179, 179)',
                },
                _a),
        },
    };
};
exports.getStyles = getStyles;
