// @fluentui/react-charting/lib-amd/components/VerticalBarChart/VerticalBarChart.styles.js
define(["require", "exports", "tslib"], function (require, exports, tslib_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    var getStyles = function (props) {
        var shouldHighlight = props.shouldHighlight, theme = props.theme;
        return {
            opacityChangeOnHover: {
                opacity: shouldHighlight ? '' : '0.1',
            },
            xAxisTicks: [],
            yAxisTicks: [
                {
                    transform: 'scaleX(-1)',
                },
            ],
            yAxisDomain: [
                {
                    transform: 'scaleX(-1)',
                },
            ],
            tooltip: tslib_1.__assign(tslib_1.__assign({}, theme.fonts.medium), { display: 'flex', flexDirection: 'column', padding: '8px', position: 'absolute', textAlign: 'center', top: '0px', background: theme.semanticColors.bodyBackground, borderRadius: '2px', pointerEvents: 'none' }),
        };
    };
    exports.getStyles = getStyles;
});
