// @fluentui/react-charting/lib-amd/components/DonutChart/DonutChart.styles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    var getStyles = function (props) {
        var className = props.className, width = props.width, height = props.height, theme = props.theme;
        return {
            root: [
                theme.fonts.medium,
                'ms-DonutChart',
                {
                    alignItems: 'center',
                    display: 'flex',
                    flexDirection: 'column',
                    width: '100%',
                    height: '100%',
                },
                className,
            ],
            chart: {
                width: width,
                height: height,
                boxSizing: 'content-box',
                overflow: 'visible',
                alignmentAdjust: 'center',
            },
            legendContainer: {
                paddingTop: '16px',
                width: width + "px",
            },
        };
    };
    exports.getStyles = getStyles;
});
