// @fluentui/react-charting/lib/components/PieChart/PieChart.base.js
import { __extends } from "tslib";
import * as React from 'react';
import { classNamesFunction } from '@fluentui/react/lib/Utilities';
import { Pie } from './Pie/Pie';
var getClassNames = classNamesFunction();
var PieChartBase = /** @class */ (function (_super) {
    __extends(PieChartBase, _super);
    function PieChartBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PieChartBase.prototype.render = function () {
        var _a = this.props, data = _a.data, width = _a.width, height = _a.height, colors = _a.colors, chartTitle = _a.chartTitle;
        var _b = this.props, theme = _b.theme, className = _b.className, styles = _b.styles, culture = _b.culture;
        this._classNames = getClassNames(styles, {
            theme: theme,
            width: width,
            height: height,
            className: className,
        });
        var radius = Math.min(width, height) / 2;
        var outerRadius = radius - 10;
        return (React.createElement("div", { className: this._classNames.root },
            this.props.chartTitle && React.createElement("p", { className: this._classNames.chartTitle }, this.props.chartTitle),
            React.createElement(Pie, { culture: culture, width: width, height: height, outerRadius: outerRadius, innerRadius: 0, data: data, colors: colors, chartTitle: chartTitle })));
    };
    PieChartBase.defaultProps = {
        data: [],
        width: 600,
        height: 350,
    };
    return PieChartBase;
}(React.Component));
export { PieChartBase };
